### PR TITLE
internal/ethapi: add status code to receipt rpc return

### PR DESCRIPTION
**Pull request by rjl493456442**

### CODE REVIEW

**Comment by mi2zha** [comment]
@rjl493456442 
hi,
Version: 1.7.0-unstable
curl http://localhost:8545 -X POST --data '{"jsonrpc":"2.0","method":"eth_getTransactionReceipt","params":["0x839517dc0f8a73f6e259745ba60badcf8016c0a7ee783ee6246ccf1157408cb3"],"id":1}'

result:

{"jsonrpc":"2.0","id":1,"result":{"blockHash":"0xf81f9f905ac74e2832fc93c50566827055b4f0df6db2e12ce092f66ec154b813","blockNumber":"0x3ffca4","contractAddress":null,"cumulativeGasUsed":"0x5dde46","from":"0x3aeda7405558a096b620315c01d3d2af383632ec","gasUsed":"0x8fb1","logs":[],"logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","root":"0xcdb37dc9126181d14d1d4846f5246750672242fa460d84b4315c0b6edc246590","to":"0xa74476443119a942de498590fe1f2454d7d4ac0d","transactionHash":"0x839517dc0f8a73f6e259745ba60badcf8016c0a7ee783ee6246ccf1157408cb3","transactionIndex":"0x34"}}

txHash "0x839517dc0f8a73f6e259745ba60badcf8016c0a7ee783ee6246ccf1157408cb3" is Bad jump destination

I use  commit b25f260
receipt.postState is no nil, why?
Thank you for your help

**Comment by rjl493456442** [comment]
@mi2zha Hey, since status field is added only after metropolish fork. So before that `postState` field is still remained and `statusCode` won't show.

**Comment by mi2zha** [comment]
@rjl493456442 
hi，
so how do you get a tx to succeed or fail

[out of gas]  [Bad jump destination] or successful

**Comment by rjl493456442** [comment]
As [Ethereum EIP658](https://github.com/ethereum/EIPs/pull/658) described, a transaction is regarded as failed only if the outermost execution failed. Any error returned (includes OutOfGas, Bad jump destination you mentioned) by the outermost call/create will make it.

**Comment by mi2zha** [comment]
@rjl493456442 

thank you for your patient answer

how did etherscan.io acquire the state of failure

debug.traceTransaction is ok? testing

**Comment by rjl493456442** [comment]
Yeah, ethereum provides lots of tools to trace the execution state of the virtual machine. You can use `debug.traceTransaction` or `evm` tool to obtain detailed vm logs.

**Comment by rjl493456442** [comment]
Actually i have added the `failed` flag into the returned result for `traceTransaction` rpc in [this pr](https://github.com/ethereum/go-ethereum/pull/15030). But it has not been merged yet.

**Comment by mi2zha** [comment]
@rjl493456442 
internal/ethapi: add status code to receipt rpc return #15042

is this ?

**Comment by mi2zha** [comment]
@rjl493456442 
 debug.traceTransaction("0x99a5e03e3cc7fd7764dfd97a357bfc1e689fddc4da2dc038ddc489662a8b6d22")

failed: true  is failed,false is successful

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1004,6 +1005,16 @@ func (s *PublicTransactionPoolAPI) GetTransactionReceipt(hash common.Hash) (map[
...
+
+	// Assign receipt status or post state.
+	if receipt.PostState != nil {
+		fields["root"] = hexutil.Bytes(receipt.PostState)
+	} else {
+		fields["statusCode"] = receiptStatusSuccessful
+		if receipt.Failed {
+			fields["statusCode"] = receiptStatusFailed
+		}
+	}
```
Comment (bas-vk) [inline internal/ethapi/api.go]: Receipt already has `statusEncoding()`. Can't we expose that one and use it instead?

Also, the EIP describes the status as a single byte, either 0 or 1. In the RPC bytes are hex encoded so this should be something like: `fields["statusCode"] = hexutil.Bytes(receipt.StatusEncoding())`. The EIP also doesn't mention that the field name is `statusCode`, that should probably be amended to the EIP.

Comment (rjl493456442) [reply internal/ethapi/api.go]: @bas-vk, I had purposed a new [EIP](https://github.com/ethereum/EIPs/pull/699) about adding a new field named `statusCode` to rpc returned. And as you mentioned, all rpc fields should be hex encoded, i think it is correct, and i will make change to satisfy this requirement.

**Comment by rjl493456442** [comment]
@bas-vk Because `receipt.statusEncoding` function is not only related with receipt status field, but also with postState field. So in my opinion, it will be more clear to assign rpc return `statusCode` with recepit's `Failed` field.

PTAL

**Comment by mi2zha** [comment]
@rjl493456442

but, there are multiple failures
[Out of gas] or  [Bad jump destination] 
how does statutscode express

**Comment by rjl493456442** [comment]
@mi2zha The statusCode could be only 1 or 0, the former represents transaction executuon successful and the latter represents failed. If you want obtain the detailed error, i think  `tracer` is a good helper.

**Comment by mi2zha** [comment]
@rjl493456442 
thank you very much

i will use debug.traceTransaction

thank you again

**Comment by bas-vk** [comment]
I don't think creating a new EIP is the way to go. The original EIP should imo specify how this new field is exposed. There is also a separate repository (https://github.com/ethereum/interfaces) that describes RPC changes.

Also I prefer to have the new field called `outcome` or `result` since it isn't really a status.

**Comment by Arachnid** [comment]
@bas-vk RPC interfaces aren't properly defined in core EIPs, IMO. There's no reason to name the status field there.

**Comment by holiman** [comment]
Updated docs: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gettransactionreceipt
LMK if it's incorrect

**Comment by rjl493456442** [comment]
@holiman The wiki has the inaccurate part. The status should be either '0x0' or '0x1'. Since all rpc responses are encoded as hex format

**Comment by karalabe** [comment]
The PR was missing the parsing part of the status field. Adding that was not really trivial because we've previously used a `Failed` field in the receipt, which gets messy when parsing into from a `status` JSON field.

After a few variations and feedback from @fjl, the consensus was to have Receipt represent the true format of the consensus type, and replace `Failed bool` to `Status uint`. My followup commit on this PR does that + the necessary code integrations to make everything work together correctly.

**Comment by karalabe** [comment]
@bas-vk PTAL

**Comment by bas-vk** [comment]
@karalabe LGTM

**Pull request merged**